# Improving Computer Vision Accuracy using Convolutions


## Without Convolution Neural Network

In [ ]:
import tensorflow as tf
mnist = tf.keras.datasets.fashion_mnist

(training_images, training_labels), (test_images, test_labels) = mnist.load_data()

4423680/4422102 [==============================] - 0s 0us/step


In [ ]:
training_images = training_images/255
test_images = test_images/255

In [ ]:
model = tf.keras.models.Sequential([
                                    tf.keras.layers.Flatten(),
                                    tf.keras.layers.Dense(512, activation=tf.nn.relu),
                                    tf.keras.layers.Dense(10, activation=tf.nn.softmax)
])

In [ ]:
model.compile(
    optimizer='adam',
    loss = 'sparse_categorical_crossentropy',
    metrics = 'accuracy'
)

model.fit(training_images, training_labels, epochs=10)
test_loss = model.evaluate(test_images, test_labels)

Epoch 1/10
1875/1875 [==============================] - 10s 5ms/step - loss: 0.3768 - accuracy: 0.8640
Epoch 2/10
1875/1875 [==============================] - 9s 5ms/step - loss: 0.3204 - accuracy: 0.8822
Epoch 3/10
1875/1875 [==============================] - 9s 5ms/step - loss: 0.3023 - accuracy: 0.8872
Epoch 4/10
1875/1875 [==============================] - 9s 5ms/step - loss: 0.2843 - accuracy: 0.8924
Epoch 5/10
1875/1875 [==============================] - 10s 5ms/step - loss: 0.2578 - accuracy: 0.9050
Epoch 6/10
1875/1875 [==============================] - 10s 5ms/step - loss: 0.2515 - accuracy: 0.9056
Epoch 7/10
1875/1875 [==============================] - 10s 5ms/step - loss: 0.2342 - accuracy: 0.9127
Epoch 8/10
1875/1875 [==============================] - 10s 5ms/step - loss: 0.2312 - accuracy: 0.9139
Epoch 9/10
1875/1875 [==============================] - 10s 5ms/step - loss: 0.2240 - accuracy: 0.9140
Epoch 10/10
313/313 [==============================] - 1s 3ms/step - loss: 0

In [ ]:
test_loss

[0.3374565541744232, 0.8881000280380249]

We have an accuracy of **88.81%. and a loss of 33.74%** Not bad but not so good either. 

## With Convolution Neural Network

In [ ]:
import tensorflow as ft
mnist = tf.keras.datasets.fashion_mnist
(training_images, training_labels), (test_images, test_labels) = mnist.load_data()

In [ ]:
training_images = training_images.reshape(60000, 28, 28, 1)
training_images = training_images/255
test_images = test_images.reshape(10000, 28, 28, 1)
test_images = test_images/255

In [ ]:
model = tf.keras.models.Sequential([
                                    tf.keras.layers.Conv2D(64, (3, 3), activation='relu', input_shape=(28, 28, 1)),
                                    tf.keras.layers.MaxPooling2D(2, 2),
                                    tf.keras.layers.Conv2D(64, (3, 3), activation='relu'),
                                    tf.keras.layers.MaxPooling2D(2, 2),
                                    tf.keras.layers.Flatten(),
                                    tf.keras.layers.Dense(512, activation=tf.nn.relu),
                                    tf.keras.layers.Dense(10, activation=tf.nn.softmax)
])

In [ ]:
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics='accuracy')
model.summary()
model.fit(training_images, training_labels, epochs=10)
my_new_accuracy = model.evaluate(test_images, test_labels)

Model: "sequential_6"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_10 (Conv2D)           (None, 26, 26, 64)        640       
_________________________________________________________________
max_pooling2d_10 (MaxPooling (None, 13, 13, 64)        0         
_________________________________________________________________
conv2d_11 (Conv2D)           (None, 11, 11, 64)        36928     
_________________________________________________________________
max_pooling2d_11 (MaxPooling (None, 5, 5, 64)          0         
_________________________________________________________________
flatten_6 (Flatten)          (None, 1600)              0         
_________________________________________________________________
dense_12 (Dense)             (None, 512)               819712    
_________________________________________________________________
dense_13 (Dense)             (None, 10)               

In [ ]:
my_new_accuracy

[0.36047619581222534, 0.9142000079154968]

With CNN our accuracy has improved to **91.42% and a loss of 36.06%**